gadgets/src/bits/boolean_input.rs

```diff
@@ -119,72 +122,50 @@ impl<F: PrimeField, CF: PrimeField> AllocGadget<Vec<F>, CF> for BooleanInputGadg
     ) -> Result<Self, SynthesisError> {
         let obj = value_gen()?;
 
-        // Step 1: obtain the bits of the F field elements (little-endian)
+        // Step 1: obtain the bits of the F field elements
         let mut src_bits = Vec::<bool>::new();
         for elem in obj.borrow().iter() {
             let mut bits = elem.to_repr().to_bits_le();
             bits.truncate(F::size_in_bits());
             for _ in bits.len()..F::size_in_bits() {
                 bits.push(false);
             }
-            bits.reverse();
-
             src_bits.append(&mut bits);
         }
 
         // Step 2: repack the bits as CF field elements
-        // Deciding how many bits can be embedded,
-        //  if CF has the same number of bits as F, but is larger,
-        //  then it is okay to put the entire field element in.
-        let capacity = if CF::size_in_bits() == F::size_in_bits() {
-            let fq = <<CF as PrimeField>::Parameters as FieldParameters>::MODULUS;
-            let fr = <<F as PrimeField>::Parameters as FieldParameters>::MODULUS;
+        let capacity = <CF::Parameters as FieldParameters>::CAPACITY;
 
-            let fq_u64: &[u64] = fq.as_ref();
-            let fr_u64: &[u64] = fr.as_ref();
+        // Step 3: allocate the CF field elements as input
+        let mut src_booleans = Vec::<Boolean>::new();
+        for (i, chunk) in src_bits.chunks(capacity as usize).enumerate() {
+            let elem = CF::from_repr(<CF as PrimeField>::BigInteger::from_bits_le(chunk)).unwrap();
 
-            let mut fq_not_smaller_than_fr = true;
-            for (left, right) in fq_u64.iter().zip(fr_u64.iter()).rev() {
-                if left < right {
-                    fq_not_smaller_than_fr = false;
-                    break;
-                }
+            let elem_gadget = FpGadget::<CF>::alloc_input(cs.ns(|| format!("alloc_elem_{}", i)), || Ok(elem))?;
 
-                if left > right {
-                    break;
-                }
-            }
+            let mut lc = LinearCombination::zero();
+            let mut coeff = CF::one();
+            let mut booleans = vec![];
```

Comment:
Ideally this vector would be created before the loop (and can be `clear`ed with every iteration) so that it doesn't get re-allocated for every `i`.